### PR TITLE
Add MFEM_VERIFY_KERNEL and MFEM_ASSERT_KERNEL.

### PR DESCRIPTION
**Pull request by YohannDudouit**
This PR adds two device-compatible macros, `MFEM_VERIFY_KERNEL` and `MFEM_ASSERT_KERNEL` that mimics the host-only macros `MFEM_VERIFY` and `MFEM_ASSERT`.
These macros can be used like the `printf` function.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#2794](https://github.com/mfem/mfem/pull/2794) | @YohannDudouit | @pazner | @artv3 + @rcarson3 + @pazner + @camierjs | 1/31/22 | 2/3/22 | 2/28/22 | |

### CODE REVIEW

**Comment by pazner** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: general/error.hpp [hunk 1]
```diff
@@ -165,18 +165,39 @@ __device__ void abort_msg(T & msg)
...
 // Abort inside a device kernel
 #if defined(__CUDA_ARCH__)
-#define MFEM_ABORT_KERNEL(msg) \
-   {                           \
-      printf(msg);             \
-      asm("trap;");            \
+#define MFEM_ABORT_KERNEL(...)                              \
+   {                                                        \
+      printf(__VA_ARGS__);                                  \
+      asm("trap;");                                         \
```
Comment (pazner) [inline general/error.hpp:171]: Is the extra whitespace needed here?

**Review by pazner — COMMENTED** [review]
Looks like a useful addition. Is it possible to add a `MFEM_VERIFY_KERNEL` and/or `MFEM_ASSERT_KERNEL` statement somewhere (either in an appropriate place in our existing kernels, or even in a unit test) to exercise this new functionality, make sure the macros work properly, etc.?

**Review by camierjs — APPROVED** [review]

**Review by rcarson3 — APPROVED** [review]
I agree with @pazner that it might be nice to see this used somewhere in the code.

**Comment by YohannDudouit** [comment]
What kind of test do you want? Should I write a GPU kernel that fails and exit with an error message?

**Comment by pazner** [comment]
It doesn't need to fail (I think that's hard to integrate into the Catch framework), but the macros should at least be expanded somewhere in the code to make sure there isn't a syntax error, etc.

**Comment by pazner** [comment]
The best would be if there is an existing kernel somewhere in the code that would benefit from these new macros. That way you don't need to introduce an artificial test.

**Comment by YohannDudouit** [comment]
@pazner I added a few `MFEM_ASSERT_KERNEL`.

**Review by pazner — APPROVED** [review]

**Comment by pazner** [comment]
Merged in `next` for testing...

**Comment by v-dobrev** [comment]
This seems to be broken with HIP. I get errors like this (in `next`):
```console
/.../mfem/linalg/sparsemat.cpp:2495:13: error: no matching constructor for initialization of 'std::ostringstream' (aka 'basic_ostringstream<char>')
            MFEM_ABORT_KERNEL("Diagonal not found in SparseMatrix::DiagScale");
            ^
/.../mfem/linalg/../general/error.hpp:182:4: note: expanded from macro 'MFEM_ABORT_KERNEL'
   MFEM_ABORT("")
   ^
/.../mfem/linalg/../general/error.hpp:100:25: note: expanded from macro 'MFEM_ABORT'
#define MFEM_ABORT(msg) _MFEM_MESSAGE("MFEM abort: " << msg, 0)
                        ^
/.../mfem/linalg/../general/error.hpp:85:26: note: expanded from macro '_MFEM_MESSAGE'
      std::ostringstream mfemMsgStream;                                 \
                         ^
/opt/rh/gcc-toolset-10/root/lib/gcc/x86_64-redhat-linux/10/../../../../include/c++/10/sstream:583:7: note: candidate constructor not viable: call to __host__ function from __device__ function
      basic_ostringstream()
      ^
```

**Comment by YohannDudouit** [comment]
@pazner I don't really understand the bug for the moment, it looks like this is not correct: `#elif defined(__HIP_DEVICE_COMPILE__)`, isn't `__HIP_DEVICE_COMPILE__` the equivalent of `__CUDA_ARCH__`?!

**Comment by pazner** [comment]
Yeah, according to the documentation I would think it would work, but it does look like it's using the host branch of `MFEM_ABORT_KERNEL`. @v-dobrev, do you have an idea what's going wrong?

**Comment by rcarson3** [comment]
@YohannDudouit so I've issues in the pass with HIP not reading `#if` type macros correctly. If you still run into trouble you might have to try something like:
```
#if defined(__CUDA_ARCH__) &&  !defined(__HIP_DEVICE_COMPILE__)
....
#endif
#if !defined(__CUDA_ARCH__) &&  defined(__HIP_DEVICE_COMPILE__)
....
#endif
#if !defined(__CUDA_ARCH__) &&  !defined(__HIP_DEVICE_COMPILE__)
...
#endif
```

Although, I am confused as what you're doing should work...

**Comment by YohannDudouit** [comment]
It seems it is using `MFEM_USE_HIP` to avoid this problem, it should work now, though we should understand why `__HIP_DEVICE_COMPILE__` does not work.

**Comment by pazner** [comment]
@nbeams, do you have any insight into why `__HIP_DEVICE_COMPILE__` isn't working as expected here?

**Comment by pazner** [comment]
Re-merged in `next` for testing...

**Comment by tzanio** [comment]
Can someone resolve the conflicts here?

**Comment by pazner** [comment]
I think the conflicts should be resolved now.

@YohannDudouit @v-dobrev, can someone confirm that this is working properly with HIP? Then I can merge it.

**Comment by YohannDudouit** [comment]
@pazner I compiled MFEM with Hip on Corona.

**Pull request merged**